File: NativeBindingDemo/Droid/MyLibrary/app/src/main/java/com/daedalus/mylibrary/Utilities.java
package com.daedalus.mylibrary;

/**
 * Created by JimmyWu on 2017/6/1.
 */

public class Utilities {
    public static String Echo(String message)
    {
        if(message.length() == 0) {
            return String.format("Dude %s, you didnt give me a message!", "bro");
        }
        return message;
    }

    public String Hello(String name)
    {
        if(name.length() == 0) {
            return String.format("Dude %s, you didnt give me a name!", "bro");
        }
        return String.format("*Waves* Hello %s! Welcome to the Xamarin binding sample!", name);
    }

    public int Add(int operandUn, int operandDeux)
    {
        return operandUn + operandDeux;
    }

    public int Multiply(int operandUn, int operandDeux)
    {
        return operandUn * operandDeux;
    }

    IUtilityCallback _callback;
    // This is an example of how to set a block function for later use.
    public void SetCallback(IUtilityCallback callback)
    {
        _callback = callback;
    }

    // This is an example of how to invoke a block function.
    public void InvokeCallback(String message)
    {
        _callback.callbackCall (message);
    }
}

File: NativeBindingDemo/iOS/MyLibraryBinding/Utilities.cs
using System;

using UIKit;
using Foundation;
using ObjCRuntime;
using CoreGraphics;

namespace MyLibraryBinding
{
	// typedef void (^UtilityCallbackDelegate)(NSString *);
	delegate void UtilityCallbackDelegate(string arg0);

	// @interface Utilities : NSObject
	[BaseType(typeof(NSObject))]
	interface Utilities
	{
		// +(NSString *)echo:(NSString *)message;
		[Static]
		[Export("echo:")]
		string Echo(string message);

		// -(NSString *)hello:(NSString *)name;
		[Export("hello:")]
		string Hello(string name);

		// -(NSInteger)add:(NSInteger)operandUn and:(NSInteger)operandDeux;
		[Export("add:and:")]
		nint Add(nint operandUn, nint operandDeux);

		// -(NSInteger)multiply:(NSInteger)operandUn and:(NSInteger)operandDeux;
		[Export("multiply:and:")]
		nint Multiply(nint operandUn, nint operandDeux);

		// -(void)setCallback:(UtilityCallbackDelegate)callback;
		[Export("setCallback:")]
		void SetCallback(UtilityCallbackDelegate callback);

		// -(void)invokeCallback:(NSString *)message;
		[Export("invokeCallback:")]
		void InvokeCallback(string message);
	}
}
